docs/machine-learning/basic-concepts-model-training-in-mldotnet.MD

```diff
@@ -0,0 +1,169 @@
+
+### Transformer chains
+
+A useful property of a transformer is that *you can phrase a sequential application of transformers as yet another transformer*:
+
+```csharp
+var fullTransformer = transformer1.Append(transformer2).Append(transformer3);
+```
+
+We utilize this property a lot in ML.NET: typically, the trained ML.NET model is a 'chain of transformers', which is, for all intents and purposes, a *transformer*. 
+
+## Data reader
+
+The data reader is ML.NET component to 'create' data: it takes an instance of `T` and returns data out of it.
+
+Here's the exact interface of `IDataReader<T>`:
+
+```c#
+public interface IDataReader<in TSource>
+{
+    IDataView Read(TSource input);
+    ISchema GetOutputSchema();
+}
+```
+
+As you can see, the reader is capable of reading data (potentially multiple times, and from different 'inputs'), but the resulting data will always have the same schema, denoted by `GetOutputSchema`.
+
+An interesting property to note is that you can create a new data reader by 'attaching' a transformer to an existing data reader. This way you can have 'reader' with transformation behavior baked in:
+
+```c#
+var newReader = reader.Append(transformer1).Append(transformer2)
+```
+
+Another similarity to transformers is that, since data is lazily evaluated, *readers are lazy*: no (or minimal) actual 'reading' happens when you call `dataReader.Read()`: only when a cursor is requested on the resulting data does the reader begin to work.
+
+## Estimator
+
+The *estimator* is an object that learns from data. The result of the learning is a *transformer*.
+Here is the interface of `IEstimator<T>`:
+
+```c#
+public interface IEstimator<out TTransformer>
+    where TTransformer : ITransformer
+{
+    TTransformer Fit(IDataView input);
+    SchemaShape GetOutputSchema(SchemaShape inputSchema);
+}
+```
+
+You can easily imagine how *a sequence of estimators can be phrased as an estimator* of its own. In ML.NET, we rely on this property to create 'learning pipelines' that chain together different estimators:
+
+```c#
+var env = new LocalEnvironment(); // Initialize the ML.NET environment.
+var estimator = new ConcatEstimator(env, "Features", "SepalLength", "SepalWidth", "PetalLength", "PetalWidth")
+    .Append(new ToKeyEstimator(env, "Label"))
+    .Append(new SdcaMultiClassTrainer(env, "Features", "Label")) // This is the actual 'machine learning algorithm'.
+    .Append(new ToValueEstimator(env, "PredictedLabel"));
+
+var endToEndModel = estimator.Fit(data); // This now contains all the transformers that were used at training.
+```
+
+One important property of estimators is that *estimators are eager, not lazy*: every call to `Fit` is causing 'learning' to happen, which is potentially a time-consuming operation.
```

Comment:
These seems like a very important design consideration:  Estimators run eagerly to create transformers, and transformers run lazily to produce transformed data.

Also, you mention that `Fit` is potentially time-consuming, but you don't give any design guidance about how to minimize the impact of that time. What should customers do to create responsive applications? Later text implies that `Fit` is performed less frequently than predictive transforms. Is that correct? Should this be stated here?

Reply:
@BillWagner The training is the time-consuming and expensive part of the process (Fit trains). Once that is stable, you can then use the model to predict values from your data. Does that help?